### PR TITLE
SLING-9796 - Add support for using Unions

### PR DESCRIPTION
**Pull request by kozmaadrian**
PR created for early review.

TODO
- [ ] Implements tests for testing the new SlingTypeProvider
- [ ] Verify SlingTypeResolverScriptWrapper
- [ ] Update Readme, describe SlingTypeResolverScriptWrapper

### CODE REVIEW

**Comment by bdelacretaz** [comment]
Could you create a https://issues.apache.org/jira/projects/SLING ticket so we can assign this to a release etc? Just a brief description of your use case there is good enough.

**Comment by kozmaadrian** [comment]
> Could you create a https://issues.apache.org/jira/projects/SLING ticket so we can assign this to a release etc? Just a brief description of your use case there is good enough.

Thanks @bdelacretaz for your input. 

I've created the following ticket https://issues.apache.org/jira/browse/SLING-9796

File: src/main/java/org/apache/sling/graphql/core/engine/SlingTypeResolverSelector.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            refs = bundleContext.getServiceReferences(SlingTypeResolver.class.getName(), filter);
+        } catch (InvalidSyntaxException ise) {
+            throw new IOException("Invalid OSGi filter syntax", ise);
+        }
+        if (refs != null) {
+            // SlingFetcher services must have a unique name
+            if (refs.length > 1) {
+                throw new IOException(String.format("Got %d services for %s, expected just one", refs.length, filter));
+            }
+            result = (SlingTypeResolver<Object>) bundleContext.getService(refs[0]);
```
Comment (raducotescu) [inline src/main/java/org/apache/sling/graphql/core/engine/SlingTypeResolverSelector.java:82]: While this follows an existing pattern in this bundle, it's not a clean approach. We increase the counter of held references for this service object and never allow them to be garbage collected (cc @bdelacretaz).

**Review by raducotescu — CHANGES_REQUESTED** [review]
There are some changes that we should do before merging. Some of those also have to happen in the current implementation of the bundle.

File: src/main/java/org/apache/sling/graphql/core/engine/SlingTypeResolverSelector.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    /**
+     * Package name prefix for fetchers which have names starting
+     * with the {#link RESERVED_NAME_PREFIX}.
+     */
+    public static final String RESERVED_PACKAGE_PREFIX = "org.apache.sling.";
+
+    private BundleContext bundleContext;
+
+    @Reference
+    private ScriptedDataFetcherProvider scriptedDataFetcherProvider;
```
[the remark this answers is not in the document]
Comment (kozmaadrian) [reply src/main/java/org/apache/sling/graphql/core/engine/SlingTypeResolverSelector.java]: Replaced with `ScriptedTypeResolverProvider, extending `getSlingTypeResolver` to call the scripted provider as fallback.

File: src/main/java/org/apache/sling/graphql/api/SlingTypeResolverEnvironment.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+     * @return the GraphQL result item
+     */
+    @Nullable
+    Object getObject();
+
+    /**
+     * @param <T> the type name
+     * @return the GraphQL Object Type
+     */
+    @Nullable <T> T getObjectType(String name);
```
[the remark this answers is not in the document]
Comment (kozmaadrian) [reply src/main/java/org/apache/sling/graphql/api/SlingTypeResolverEnvironment.java]: Removed `@Nullable`.

File: src/main/java/org/apache/sling/graphql/core/engine/ScriptedTypeResolverProvider.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+ */
+@Component(service = ScriptedTypeResolverProvider.class)
+public class ScriptedTypeResolverProvider {
+    public static final String SCRIPT_NAME = "resolver";
+    public static final String FAKE_RESOURCE_TYPE_PREFIX = "graphql/resolver/";
+
+    @Reference
+    private ServletResolver servletResolver;
+
+    private static class FakeResource extends AbstractResource {
```
[the remark this answers is not in the document]
Comment (bdelacretaz) [reply src/main/java/org/apache/sling/graphql/core/engine/ScriptedTypeResolverProvider.java:48]: SyntheticResource should work indeed, I agree

File: src/main/java/org/apache/sling/graphql/core/engine/ScriptedTypeResolverProvider.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+import javax.servlet.Servlet;
+
+/**
+ * Resolves a SlingTypeResolver name to a SlingScript if available.
+ */
+@Component(service = ScriptedTypeResolverProvider.class)
+public class ScriptedTypeResolverProvider {
```
Comment (bdelacretaz) [inline src/main/java/org/apache/sling/graphql/core/engine/ScriptedTypeResolverProvider.java:41]: Do we really need a scripted type provider? I understand it exists here by analogy with the scripted SlingDataFetcher but I'm not sure about the use cases for it.

File: src/main/java/org/apache/sling/graphql/core/engine/SlingTypeResolverSelector.java (hunk 1 above)
Comment (bdelacretaz) [reply src/main/java/org/apache/sling/graphql/core/engine/SlingTypeResolverSelector.java:82]: I suppose we can factor out the logic which is similar for this and for the SlingDataFetcher? And clean it up in a single place then.

File: src/main/java/org/apache/sling/graphql/core/engine/ScriptedTypeResolverProvider.java (hunk 5 above)
Comment (kozmaadrian) [reply src/main/java/org/apache/sling/graphql/core/engine/ScriptedTypeResolverProvider.java:41]: Actually, I tried to follow the concept used for DataFetchers.

File: src/main/java/org/apache/sling/graphql/core/engine/SlingTypeResolverSelector.java (hunk 1 above)
Comment (raducotescu) [reply src/main/java/org/apache/sling/graphql/core/engine/SlingTypeResolverSelector.java:82]: Yes - I opened https://issues.apache.org/jira/browse/SLING-9800 which besides the improvement regarding query execution should also optimised the interaction with the service registry.

File: src/main/java/org/apache/sling/graphql/core/engine/ScriptedTypeResolverProvider.java (hunk 5 above)
Comment (raducotescu) [reply src/main/java/org/apache/sling/graphql/core/engine/ScriptedTypeResolverProvider.java:41]: I talked offline with @bdelacretaz and we don't think that a `ScriptedTypeResolverProvider` makes sense for the following reasons:
* it's hard to test by developers, requiring ITs
* it's kind of cumbersome to write one

**Comment by raducotescu** [comment]
I've applied a slightly modified patch in https://github.com/apache/sling-org-apache-sling-graphql-core/pull/8, since my change from https://github.com/apache/sling-org-apache-sling-graphql-core/pull/7 made this PR unmergeable. Let's continue the discussion there.

**Pull request closed without merging**